cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/RemoteAppStateHandler.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.runtime;
+
+import com.google.inject.Inject;
+import io.cdap.cdap.api.app.AppStateException;
+import io.cdap.cdap.api.app.AppStateReader;
+import io.cdap.cdap.api.app.AppStateWriter;
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.common.ServiceUnavailableException;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.http.DefaultHttpRequestConfig;
+import io.cdap.cdap.common.internal.remote.RemoteClient;
+import io.cdap.cdap.common.internal.remote.RemoteClientFactory;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpRequest;
+import io.cdap.common.http.HttpResponse;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.nio.ByteBuffer;
+
+/**
+ * Remote handler for application state
+ */
+public class RemoteAppStateHandler implements AppStateReader, AppStateWriter {
```

Comment:
Don't call a class `Handler` if it is not a HTTP handler. Maybe call this `RemoteAppStateStore`. Also, do we need to have two different interfaces for read/write?

Reply:
Currently we do not have requirement to read state without write. Is this a possibility in the future for any plugin? Otherwise can merge them into one interface.

Reply:
Combined to one after discussing offline.